File: src/main/python/GrammarFuzzer.py
import random
import re
from typing import List, Tuple

# Type alias for clarity
ProbabilisticGrammar = dict[str, list[tuple[str, float]]]


class ProbabilisticGrammarFuzzer:
    def __init__(self, grammar: ProbabilisticGrammar,
                 random_instance: random.Random):
        self.grammar = grammar
        self.random = random_instance

    def nonterminals(self, expansion: str) -> List[str]:
        return re.findall(r"\[[^\[\]]+\]", expansion)

    def select_probabilistic_expansion(self,
                                       expansions: List[Tuple[str, float]]) -> str:
        total_weight = sum(weight for _, weight in expansions)
        random_point = self.random.uniform(0, total_weight)

        cumulative_weight = 0.0
        for expansion, weight in expansions:
            cumulative_weight += weight
            if random_point <= cumulative_weight:
                return expansion
        return expansions[-1][
            0]  # Fallback to last expansion in case of rounding errors

    def fuzz(self, start_symbol: str = "[start]", max_num_of_nonterminals: int = 10,
             max_expansion_trials: int = 100, verbose: bool = False) -> str:
        term = start_symbol
        expansion_trials = 0

        while self.nonterminals(term):
            symbol_to_expand = random.choice(self.nonterminals(term))
            expansions = self.grammar.get(symbol_to_expand)

            if expansions is None:
                raise ValueError(f"Unknown symbol {symbol_to_expand}")

            expansion = self.select_probabilistic_expansion(expansions)
            new_term = term.replace(symbol_to_expand, expansion, 1)

            if len(self.nonterminals(new_term)) < max_num_of_nonterminals:
                term = new_term
                if verbose:
                    print(f"{symbol_to_expand} -> {expansion} : {term}")
                expansion_trials = 0  # Reset expansion trials on success
            else:
                expansion_trials += 1
                if expansion_trials >= max_expansion_trials:
                    break

        return term

File: src/main/kotlin/me/markoutte/joker/jsoupParse/GrammarFuzzer.kt
package me.markoutte.joker.jsoupParse

import java.nio.charset.Charset
import kotlin.random.Random

typealias ProbabilisticGrammar = Map<String, List<Pair<String, Double>>>

class ProbabilisticGrammarFuzzer(
    val grammar: ProbabilisticGrammar,
    val random: Random
) {

    private fun nonterminals(expansion: String): List<String> {
        return Regex("\\[[^\\[\\]]+\\]").findAll(expansion).map { it.value }.toList()
    }

    private fun selectProbabilisticExpansion(expansions: List<Pair<String, Double>>): String {
        val totalWeight = expansions.sumOf { it.second }
        val randomPoint = random.nextDouble(totalWeight)

        var cumulativeWeight = 0.0
        for (expansion in expansions) {
            cumulativeWeight += expansion.second
            if (randomPoint <= cumulativeWeight) {
                return expansion.first
            }
        }
        return expansions.last().first
    }

    fun fuzz(
        startSymbol: String = "[start]",
        maxNumOfNonterminals: Int,
        maxExpansionTrials: Int,
        charset: Charset = Charsets.UTF_8,
        verbose: Boolean = false
    ): ByteArray {
        var term = startSymbol
        var expansionTrials = 0

        while (nonterminals(term).isNotEmpty()) {
            val symbolToExpand = nonterminals(term).random()
            val expansions = grammar[symbolToExpand]
                ?: throw IllegalArgumentException("Unknown symbol $symbolToExpand")

            val expansion = selectProbabilisticExpansion(expansions)
            val newTerm = term.replaceFirst(symbolToExpand, expansion)

            if (nonterminals(newTerm).size < maxNumOfNonterminals) {
                term = newTerm
                if (verbose) {
                    println("$symbolToExpand -> $expansion : $term")
                }
                expansionTrials = 0
            } else {
                expansionTrials++
                if (expansionTrials >= maxExpansionTrials) {
                    break
                }
            }
        }
        return term.toByteArray(charset)
    }
}